pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/MVScanDocIdIterator.java

```diff
@@ -73,17 +74,27 @@ public int advance(int targetDocId) {
 
   @Override
   public MutableRoaringBitmap applyAnd(ImmutableRoaringBitmap docIds) {
-    MutableRoaringBitmap result = new MutableRoaringBitmap();
-    IntIterator docIdIterator = docIds.getIntIterator();
-    int nextDocId;
-    while (docIdIterator.hasNext() && (nextDocId = docIdIterator.next()) < _numDocs) {
-      int length = _reader.getDictIdMV(nextDocId, _dictIdBuffer, _readerContext);
-      _numEntriesScanned += length;
-      if (_predicateEvaluator.applyMV(_dictIdBuffer, length)) {
-        result.add(nextDocId);
+    if (docIds.isEmpty()) {
+      return new MutableRoaringBitmap();
+    }
+    RoaringBitmapWriter<MutableRoaringBitmap> result = RoaringBitmapWriter.bufferWriter()
+        .expectedRange(docIds.first(), docIds.last()).runCompress(false).get();
+    BatchIterator docIdIterator = docIds.getBatchIterator();
+    int[] buffer = new int[OPTIMAL_ITERATOR_BATCH_SIZE];
+    while (docIdIterator.hasNext()) {
+      int limit = docIdIterator.nextBatch(buffer);
+      for (int i = 0; i < limit; i++) {
+        int nextDocId = buffer[i];
+        if (nextDocId < _numDocs) {
```

Comment:
(Optional and not introduced in this PR) I think this if check is redundant. We need this check for the old star-tree, which is already removed and no longer supported.

Reply:
good suggestion, done.